intake_dal/dal_online.py

```diff
@@ -0,0 +1,188 @@
+from urllib.parse import urlparse, ParseResult, urldefrag  # noqa: F401
+import base64
+import io
+import json
+import urllib.parse
+from collections import Iterable
+from typing import Dict, Optional
+
+import numpy as np
+import pandas as pd
+import pandavro
+import pkg_resources
+import requests
+from intake import DataSource, Schema
+
+
+class DalOnlineSource(DataSource):
+    """
+    Interfaces with Online FS
+    This Intake source is to be used in conjunction with intake-dal
+    """
+    container = "dataframe"
+    partition_access = False
+    name = "dal-online"
+    version = pkg_resources.get_distribution("intake-dal").version
+
+    def __init__(self, urlpath="", key=None, storage_options=None, metadata=None):
+        """
+            Fetches from Online FS
+        Args:
+            urlpath: the host path
+            key: key value
+            key_name: key name
+            storage_options: optional storage options to send along
+            metadata:
+        """
+
+        self._urlpath = urlpath
+        parse_result = urlparse(urlpath)  # type: ParseResult
+        (self._url, _) = urldefrag(urlpath)
+        self._key_name = parse_result.fragment
+        assert self._key_name != '', "key_name expected in URL fragment"
+        self._key_value = key
+        self._canonical_name = None  # _get_schema() sets this
+        super(DalOnlineSource, self).__init__(metadata=metadata)
+
+    def _get_schema(self):
+        if self._canonical_name is None:
+            self._canonical_name = self.metadata['canonical_name']  # an intake-dal feature
+            # TODO(talebz): Getting avro schema should be promoted to Intake
+            self._avro_schema = _get_avro(self, self._canonical_name)
+            self._schema_dtypes = _avro_to_dtype(self._avro_schema)
+            self._dtypes = {k: str(v) for (k, v) in self._schema_dtypes.items()}
+
+        return Schema(
+            datashape=None,
+            dtype=self._dtypes,
+            shape=(None, len(self._dtypes)),
+            npartitions=1,  # This data is not partitioned, so there is only one partition
+            extra_metadata={
+                'canonical_name': self._canonical_name,
+                'urlpath': self._urlpath
+            },
+        )
+
+    def write(self, df: pd.DataFrame):
+        self._get_schema()
+        avro_str = serialize_panda_df_to_str(df, self._avro_schema)
+        self._http_put_avro_data_set(
+            {'data_set_name': self._canonical_name, 'key_name': self._key_name, 'avro_rows': avro_str}
+        )
+
+    def _get_partition(self, _) -> pd.DataFrame:
```

Comment:
function name seems to be completely different from what the body indicates (get partition vs deserialize an avro string)?

Reply:
This function name is implementing an Intake base class abstract method.